ui/app/components/app/metamask-template-renderer/metamask-template-renderer.js

```diff
@@ -0,0 +1,67 @@
+import React, { memo } from 'react'
+import PropTypes from 'prop-types'
+import { isEqual } from 'lodash'
+import { safeComponentList } from './safe-component-list'
+
+const MetaMaskTemplateRenderer = ({ sections }) => {
+  if (!sections) {
+    return null
+  }
+  if (typeof sections === 'string') {
+    return sections
+  }
+  return (
+    <>
+      {sections.reduce((allChildren, child, index) => {
+        // React can render strings directly, so push them into the accumulator
+        if (typeof child === 'string') {
+          allChildren.push(child)
+        } else if (child) {
+          // The other option is one of our Sections, which contains
+          // element, children, and props.
+          const { element, children, props } = child
+          const Element = safeComponentList[element]
+          if (!Element) {
+            console.warn(
+              `${element} is not in the safe component list and will not be rendered`,
+            )
+          }
+          const childrenOrNull =
+            Element && children ? (
+              <MetaMaskTemplateRenderer sections={children} />
+            ) : null
+          if (Element) {
+            allChildren.push(
+              <Element key={`${element}${index + 0}`} {...props}>
```

Comment:
Should we have a default `key`? :thinking: This should probably be provided to the template renderer, to ensure that distinct elements are treated as distinct by React. To do otherwise risks breaking in situations where similar components get moved around, and that moving has a side-effect (e.g. CSS transitions). Effectively we're silencing the React warning here.

Reply:
I get why you did it though. Requiring a `key` for every entry is demanding.

We could eliminate the `key` requirement in the single-element case by returning it directly, rather than in an array. It'd require another condition in the renderer for cases where `sections` is a single element or an array with one entry.

Reply:
I added object-hash, which I was planning on using as part of the transaction -> activity refactor and grouping by intent. This allows me to generate a key based on the shape of sections, the child we are rendering, and the current index. I strongly prefer to not make key something that is required in the object entries if we can avoid it.

Reply:
This again serves only to silence the warning, and defeats the purpose of the `key`: to distinguish the identity of components as props change. This is functionally identical to what this was before, isn't it?

Reply:
Actually, I'd stake some eth on this being *worse* than just using the element name and the index. What I did will actually change the key anytime props change anywhere in the tree because the props were included in the hash. I'm not sure what I was thinking, keys are meant to provide a stable identifier for the element. It should be the same even when props change.

I still really feel like adding a key to every entry to be tedious.

>1. the list and items are static–they are not computed and do not change;
>2. the items in the list have no ids;
>3. the list is never reordered or filtered.
>
>When all of them are met, you may safely use the index as a key.
Source: [index-as-akey-is-an-anti-pattern](https://robinpokorny.medium.com/index-as-a-key-is-an-anti-pattern-e0349aece318)

_the list and items are static-they are not computed and do not change_	
with the current implementation, the `sections` object only changes when the confirmation itself changes. In these cases, we're dealing with, potentially, an entirely different tree of components. They are computed, so it doesn't meet the full requirement, but we aren't dealing with a list of items in a `<li>`.

_the items in the list have no ids_
I don't see a way of generating a stable id for these objects, but we could ask the dev to id the section manually. 

_the list is never reordered or filtered_
We can't make any strong assurances on this. In the future, we may want to render a sortable list component with the template language, which would place us in violation of this rule as well.

I'm going to work on implementing your suggestion, @Gudahtt

Reply:
Maybe we could add an optional prop to the renderer that declares something as static, meeting all three of those requirements? Then we could auto-generate keys in the manner you did previously.

Reply:
I updated this to allow an object as valid children, indicating a single node.

To recap possible values for `sections` and `sections.children` are now:
1. a string
2. a object with the shape of `sections`
3. an array of objects with the shape of `sections`


when dealing with an object or string, we don't require key. When dealing with an array we do. 

check the latest commit to see how the story/code changed. I think this is the safest approach. Optimizations around key generation can come later if it becomes burdensome.

Reply:
@Gudahtt i just rewrote that commit a bit, its now much simpler

Reply:
One more optimization to reduce recursion calls where appropriate. 

![image](https://user-images.githubusercontent.com/4448075/106635333-5437ef00-6546-11eb-8f80-c07e5c9a799f.png)

highlighting the key being attached to various places.